core-processor/src/processing.rs

```diff
@@ -139,25 +139,54 @@ where
     });
 
     match exec_result {
-        Ok(res) => Ok(match res.kind {
-            DispatchResultKind::Trap(reason) => process_execution_error(
-                res.dispatch,
-                program_id,
-                res.gas_amount.burned(),
-                res.system_reservation_context,
-                ActorExecutionErrorReplyReason::Trap(reason),
-            ),
-            DispatchResultKind::Success => process_success(Success, res),
-            DispatchResultKind::Wait(duration, ref waited_type) => {
-                process_success(Wait(duration, waited_type.clone()), res)
+        Ok(res) => {
+            match res.kind {
+                DispatchResultKind::Success
+                | DispatchResultKind::Wait(_, _)
+                | DispatchResultKind::Exit(_) => {
+                    // assert that after processing the initial reservation is less or equal to the current one.
+                    // during execution reservation amount might increase due to `system_reserve_gas` calls
+                    // thus making initial reservation less than current one.
+                    debug_assert!(system_reservation_ctx
```

Comment:
both these checks are aimed to verify system reservations
1. these checks aren't 100% correct
2. you need to check initial task's idea: amount of common reservations (not the amount of gas in system reservation)